### PR TITLE
add check of certificate existence

### PR DESCRIPTION
**Pull request by spetrunin**
add checking that provided cert is matching certificate in the key

## Description

Extend key auth checks with:
- [x] checking that certificate is present in a system
- [x] checking that certificate is matching certificate from the key
- [x] check that key could not be used without cert

## Related Issue

https://tyktech.atlassian.net/browse/TT-177 -> https://github.com/TykTechnologies/tyk-analytics/issues/2023
https://tyktech.atlassian.net/browse/TT-189 -> https://github.com/TykTechnologies/tyk-analytics/issues/1634

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [ ] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
  - [ ] If new config option added, ensure that it can be set via ENV variable
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] When updating library version must provide reason/explanation for this update.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Check your code additions will not fail linting checks:
  - [ ] `go fmt -s`
  - [ ] `go vet`

### CODE REVIEW

**Comment by sredxny** [comment]
Note: the check  _check that key could not be used without cert_ is validated in dashboard side, as we are deleting and creating a new key, then only place where we know the old and new key is dashboard (there we can check if the old key was using cert then the new one will require a cert as well)...done in dashboard pr

**Comment by sredxny** [comment]
Added tests to cover new validations, updated existing tests. Now we check via tests that we cannot use keys with erased certificates

File: gateway/api.go [hunk 1]
```diff
@@ -350,6 +350,22 @@ func handleAddOrUpdate(keyName string, r *http.Request, isHashed bool) (interfac
 		}
 		originalKey = key.Clone()
 
+		isCertificateChanged := newSession.Certificate != originalKey.Certificate
+		if isCertificateChanged{
+			if newSession.Certificate == ""{
```
Comment (matiasinsaurralde) [inline gateway/api.go:355]: Should we check for `newSession.Certificate == ""` earlier?

**Review by matiasinsaurralde — APPROVED** [review]
In general looks good, have added a question though

**Review by tbuchaillot — APPROVED** [review]
LGTM

**Review by matiasinsaurralde — APPROVED** [review]

**Pull request merged**